### PR TITLE
Bug correction: Don't allow empty html or whitespaces in content

### PR DESCRIPTION
**Pull request by xaviqv**
There are currently some errors regarding empty content in several fields:

**Subject**
- Whitespaces in the beginning or the end of are treated as valid characters.

**Ticket description, Ticket comment**
- Whitespaces in the beginning or the end are treated as valid characters.
- Not visible HTML (for example empty rows) are treated as valid characters.
- Validation required lenght is based on html input

Corrections applied:
- Input trimming (both for subject and content)
- Content validation now uses the purified content without html tags. This prevents also accepting content that includes a screenshot without any text before or after it.

### CODE REVIEW

File: src/Traits/Purifiable.php [hunk 1]
```diff
@@ -8,17 +8,17 @@
...
     /**
-     * Updates the content and html attribute of the given model.
+     * Returns an array with both filtered and excluded html.
      *
      * @param string $rawHtml
      *
-     * @return \Illuminate\Database\Eloquent\Model $this
+     * @return array
      */
-    public function setPurifiedContent($rawHtml)
```
Comment (thekordy) [inline src/Traits/Purifiable.php:17]: Don't break the API please

File: src/Controllers/CommentsController.php [hunk 2]
```diff
@@ -43,14 +46,18 @@ public function create()
      */
     public function store(Request $request)
     {
+        $a_content = $this->purifyHtml($request->get('content'));
```
Comment (thekordy) [inline src/Controllers/CommentsController.php:49]: please use meaningful names

File: src/Controllers/CommentsController.php [hunk 3]
```diff
@@ -43,14 +46,18 @@ public function create()
...
+        $request->merge(['content'=>$a_content['content']]);
+
         $this->validate($request, [
             'ticket_id'   => 'required|exists:ticketit,id',
             'content'     => 'required|min:6',
         ]);
 
         $comment = new Models\Comment();
 
-        $comment->setPurifiedContent($request->get('content'));
```
Comment (thekordy) [inline src/Controllers/CommentsController.php:53]: Why  not that API?

File: src/Controllers/TicketsController.php [hunk 4]
```diff
@@ -172,6 +175,12 @@ public function create()
      */
     public function store(Request $request)
     {
+        $a_content = $this->purifyHtml($request->get('content'));
```
Comment (thekordy) [inline src/Controllers/TicketsController.php:178]: meaningful names please

File: src/Controllers/TicketsController.php [hunk 5]
```diff
@@ -183,7 +192,8 @@ public function store(Request $request)
 
         $ticket->subject = $request->subject;
 
-        $ticket->setPurifiedContent($request->get('content'));
```
Comment (thekordy) [inline src/Controllers/TicketsController.php:186]: Why not use the current purify trait?

File: src/Controllers/TicketsController.php [hunk 6]
```diff
@@ -196,7 +206,7 @@ public function store(Request $request)
 
         session()->flash('status', trans('ticketit::lang.the-ticket-has-been-created'));
 
-        return redirect()->action('\Kordy\Ticketit\Controllers\TicketsController@index');
+        //return redirect()->action('\Kordy\Ticketit\Controllers\TicketsController@index');
```
Comment (thekordy) [inline src/Controllers/TicketsController.php]: why?

**Review by thekordy — COMMENTED** [review]
I think a minor trim to the setPurifiedContent method would solve it all

File: src/Traits/Purifiable.php [hunk 7]
```diff
@@ -8,17 +8,17 @@
...
+    public function purifyHtml($rawHtml)
     {
-        $this->content = Purifier::clean($rawHtml, ['HTML.Allowed' => '']);
-        $this->html = Purifier::clean($rawHtml, Setting::grab('purifier_config'));
```
Comment (thekordy) [inline src/Traits/Purifiable.php:20]: Wouldn't a trim here be enough?

**Review by thekordy — CHANGES_REQUESTED** [review]

File: src/Controllers/CommentsController.php (hunk 3 above)
Comment (xaviqv) [reply src/Controllers/CommentsController.php:53]: The first and simplest way to filter before validate() using the Purifiable trait I found was to add additional methods to it to have my needed string outputs. Then I though that for performance reasons, having already the purified strings before validate(), It was not needed to run it again when creating the ticket instance. I saw also that the Purifiable trait was only used in the Tickets and Comments controllers. If I had seen that used anywhere else on the project I think I have then done the first approach.

Anyway, my only objective with this PR is to correct the form functionality, so please tell me what do you think about this.

File: src/Controllers/TicketsController.php (hunk 4 above)
Comment (xaviqv) [reply src/Controllers/TicketsController.php:178]: Maybe would be better something like "getPurifiedHtml()?"

File: src/Controllers/TicketsController.php (hunk 6 above)
Comment (xaviqv) [reply src/Controllers/TicketsController.php]: I'm sorry. It was a modification for testing only!

File: src/Traits/Purifiable.php (hunk 7 above)
Comment (xaviqv) [reply src/Traits/Purifiable.php:20]: In my testings I started with that simple trim(). I don't know why, but It didn't make any effect on output, both for subject or content. Then I found out the parameter regarding to new rows, tabs... etc " \t\n\r\0\x0B". It also didnt make effect on the output. I though it was that somewhere in the code, the spaces in the form fields are identified or converted to something else.

Finally, the difference was adding "chr(0xC2).chr(0xA0)" that I found on [one php.net trim() comment](http://php.net/manual/en/function.trim.php#98812). I keeped also the " \t\n\r\0\x0B" string to clean the input in other cases that I didn't foresee.

**Comment by thekordy** [comment]
not ready for production, sorry.

**Comment by roni-estein** [comment]
Thank you!
Spark by Readdle

**Pull request closed without merging**